src/main.cc

```diff
@@ -214,6 +231,145 @@ int main(int argc, char** argv, char** envp)
+                exit(EXIT_FAILURE);
+            }
+
+            // set home according to /etc/passwd entry
+            if (!home.has_value()) {
+                home = user_id->pw_dir;
+            }
+
+            // we need to be euid root to become requested user/group
+            if (actual_euid != 0) {
+                log_error("Need to be root to change user.");
+                exit(EXIT_FAILURE);
+            }
+
+            // set all uids, gids and add. groups
+// mac os x does this differently, setgid and setuid are basically doing the same
+// as setresuid and setresgid on linux: setting all of real{u,g}id, effective{u,g}id and saved-set{u,g}id
+#ifdef __APPLE__
+            // set group-ids, then add. groups, last user-ids, all need to succeed
+            if (0 != setgid(user_id->pw_gid) || 0 != initgroups(user_id->pw_name, user_id->pw_gid) || 0 != setuid(user_id->pw_uid)) {
+#else
+            if (0 != setresgid(user_id->pw_gid, user_id->pw_gid, user_id->pw_gid) || 0 != initgroups(user_id->pw_name, user_id->pw_gid) || 0 != setresuid(user_id->pw_uid, user_id->pw_uid, user_id->pw_uid)) {
+#endif
+                log_error("Unable to change user.");
+                exit(EXIT_FAILURE);
+            }
+            log_info("Dropped to User: {}", user->c_str());
+        }
+
+        // are we requested to daemonize?
+        bool daemon = opts["daemon"].as<bool>();
+        if (daemon) {
+            pid_t pid;
+
+            // fork
+            pid = fork();
+            if (pid < 0) {
+                log_error("Unable to fork.");
+                exit(EXIT_FAILURE);
+            }
+
+            // terminate parent
+            if (pid > 0)
+                exit(EXIT_SUCCESS);
+
+            // become session leader
+            if (setsid() < 0) {
+                log_error("Unable to setsid.");
+                exit(EXIT_FAILURE);
+            }
+
+            // second fork
+            pid = fork();
+            if (pid < 0) {
+                log_error("Unable to fork.");
+                exit(EXIT_FAILURE);
+            }
+
+            // terminate parent
+            if (pid > 0)
+                exit(EXIT_SUCCESS);
+
+            // set new file permissions
+            umask(0);
+
+            // change dir to /
+            if (chdir("/")) {
+                log_error("Unable to chdir to root dir.");
+                exit(EXIT_FAILURE);
+            }
+            // close open filedescriptors belonging to a tty
+            for (int fd = sysconf(_SC_OPEN_MAX); fd >= 0; fd--) {
+                if (isatty(fd))
+                    close(fd);
+            }
+            log_info("Daemonized.");
+        }
+
+        // are we requested to write a pidfile ?
+        std::optional<std::string> pidfile;
+        if (opts.count("pidfile") > 0) {
+            pidfile = opts["pidfile"].as<std::string>();
+
+            // exit if the pidfile already exists
+            struct stat* s = NULL;
```

Comment:
You'd better make it an object `struct stat statbuf;` and call `stat()` with `&statbuf`

Reply:
Good point, thanks!